caffe2/core/tensor.h

```diff
@@ -509,17 +509,48 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
    * using it. If a Deleter object is passed in, when this tensor is reallocated
    * or freed, the deleter function is going to be called.
    */
-  template <typename T, typename Deleter = MemoryDeleter>
-  void ShareExternalPointer(T* src, size_t capacity = 0, Deleter d = nullptr) {
-    ShareExternalPointer(src, TypeMeta::Make<T>(), capacity, d);
+  template <typename T>
+  void
+  ShareExternalPointer(T* src, size_t capacity = 0, MemoryDeleter d = nullptr) {
+    ShareExternalPointer((void*)src, TypeMeta::Make<T>(), capacity, d);
+  }
+
+  template <typename T>
+  void ShareExternalPointer(at::DataPtr&& data_ptr, size_t capacity = 0) {
+    ShareExternalPointer(std::move(data_ptr), TypeMeta::Make<T>(), capacity);
   }
 
-  template <typename Deleter = MemoryDeleter>
   void ShareExternalPointer(
       void* src,
       const TypeMeta& data_type,
       size_t capacity = 0,
-      Deleter d = nullptr) {
+      MemoryDeleter d = nullptr) {
+    CAFFE_ENFORCE_WITH_CALLER(
+        data_type.id() != TypeIdentifier::uninitialized(),
+        "To share with a raw external pointer you need to pass in an "
+        "initialized data_type(TypeMeta).");
+    if (!capacity) {
+      capacity = numel_ * data_type.itemsize();
+    }
+    at::DataPtr data_ptr = at::DataPtr(src, nullptr, d, GetDeviceType());
```

Comment:
The deleter is intended to be called on `src`, so you need to set src as both the data and the context pointer, e.g., `src, src, d, GetDeviceType()`